pkg/license/aggregator.go

```diff
@@ -0,0 +1,208 @@
+	var kbList kbv1.KibanaList
+	err := a.client.List(&kbList)
+	if err != nil {
+		return resource.Quantity{}, errors.Wrap(err, "failed to aggregate Kibana memory")
+	}
+
+	var total resource.Quantity
+	for _, kb := range kbList.Items {
+		mem, err := containerMemLimits(
+			kb.Spec.PodTemplate.Spec.Containers,
+			kbv1.KibanaContainerName,
+			kbconfig.EnvNodeOpts, memFromNodeOptions,
+			kbpod.DefaultMemoryLimits,
+		)
+		if err != nil {
+			return resource.Quantity{}, errors.Wrap(err, "failed to aggregate Kibana memory")
+		}
+
+		total.Add(multiply(mem, kb.Spec.Count))
+		log.V(1).Info("Collecting", "kibana_name", kb.Name, "memory", mem.String(), "count", kb.Spec.Count)
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateApmServerMemory() (resource.Quantity, error) {
+	var asList apmv1.ApmServerList
+	err := a.client.List(&asList)
+	if err != nil {
+		return resource.Quantity{}, errors.Wrap(err, "failed to aggregate APM Server memory")
+	}
+
+	var total resource.Quantity
+	for _, as := range asList.Items {
+		mem, err := containerMemLimits(
+			as.Spec.PodTemplate.Spec.Containers,
+			apmv1.ApmServerContainerName,
+			"", nil, // no fallback with limits defined in an env var
+			apmserver.DefaultMemoryLimits,
+		)
+		if err != nil {
+			return resource.Quantity{}, errors.Wrap(err, "failed to aggregate APM Server memory")
+		}
+
+		total.Add(multiply(mem, as.Spec.Count))
+		log.V(1).Info("Collecting", "as_name", as.Name, "memory", mem.String(), "count", as.Spec.Count)
+	}
+
+	return total, nil
+}
+
+// containerMemLimits reads the container memory limits from the resource specification with fallback
+// on the environment variable and on the default limits
+func containerMemLimits(
+	containers []corev1.Container,
+	containerName string,
+	envVarName string,
+	envLookup func(envVar string) (resource.Quantity, error),
+	defaultLimit resource.Quantity,
+) (resource.Quantity, error) {
+	var mem resource.Quantity
+	for _, container := range containers {
+		if container.Name == containerName {
+			mem = *container.Resources.Limits.Memory()
+
+			// if not, maybe fallback to the environment variable
+			if envLookup != nil && mem.IsZero() {
+				for _, envVar := range container.Env {
+					if envVar.Name == envVarName {
+						var err error
+						mem, err = envLookup(envVar.Value)
+						if err != nil {
+							return resource.Quantity{}, err
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// if not, fallback to the default limits
+	if mem.IsZero() {
+		mem = defaultLimit
+	}
+
+	return mem, nil
+}
+
+// maxHeapSizePattern is the pattern to extract the max Java heap size (-Xmx<size>[g|G|m|M|k|K] in binary units)
+var maxHeapSizeRe = regexp.MustCompile(`-Xmx([0-9]+)([gGmMkK]?)(?:\s.+|$)`)
+
+// memFromJavaOpts extracts the maximum Java heap size from a Java options string, multiplies the value by 2
+// and converts it to a resource.Quantity
+func memFromJavaOpts(javaOpts string) (resource.Quantity, error) {
```

Comment:
I'd add a bit more context to the "multiply by 2" thing, explaining how giving twice the JVM memory to the container is a common thing people do